go/data-source-sql/countries/countries.go

```diff
@@ -0,0 +1,88 @@
+package countries
+
+// Access to a database of country names.
+
+import (
+	"fmt"
+	"log"
+	"database/sql"
+	"strings"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/signavio/connector-examples/go/data-source-sql/configuration"
+)
+
+var databaseDriver = "mysql"
+var databaseUrl = "cldr:cldr@tcp(127.0.0.1:3306)/cldr"
+
+type Countries struct {
+  db *sql.DB
+}
+
+// JSON type required by the connector API.
+type Country struct {
+	Code string `json:"id"`
+	Name string `json:"name"`
+}
+
+type CountryCodeUndefined string
+
+func (code CountryCodeUndefined) Error() string {
+	return fmt.Sprintf("No country defined for code ‘%v’", string(code))
+}
+
+// Connect to the database.
+func NewCountries(configuration configuration.Configuration) (*Countries) {
+  countries := Countries{}
+  countries.db = Database(configuration)
+  return &countries
+}
+
+// Connects to the database and tests the connection
+func Database(configuration configuration.Configuration) *sql.DB {
+	db, err := sql.Open(configuration.DatabaseDriver, configuration.DatabaseUrl)
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+	if err = db.Ping(); err != nil {
+		log.Fatal(err.Error())
+	}
+	log.Print("Connected to database")
+	return db
+}
+
+// Finds a country by code.
+func (c Countries) FindOne(code string) (country Country, err error) {
+	if c.db == nil {
+	  log.Fatal("No database")
+	}
+	err = c.db.QueryRow("select code, name from countries where code = ?", code).Scan(&country.Code, &country.Name)
+	if err != nil {
+		if (err.Error() == "sql: no rows in result set") {
+			err = CountryCodeUndefined(code)
+		}
```

Comment:
Instead you could do this:

```go
errNoRows, ok := err.(sql.ErrNoRows)
if ok {
  return c, CountryCodeUndefined(code)
}

```

Reply:
Much nicer! I’ll add a commit for that.

Reply:
For what it’s worth, this also seems to work (see in the [sql_test.go](https://golang.org/src/database/sql/sql_test.go?h=ErrNoRows#L1786) source.

```go
		if err == sql.ErrNoRows {
		  return Country{}, CountryCodeUndefined(code)
		}
```

Reply:
Nice find! That is definitely the way to do it! :+1: